src/gamepad/mapping.vala

```diff
@@ -0,0 +1,236 @@
+// This file is part of GNOME Games. License: GPLv3
+
+private class LibGamepad.DPad : Object {
+	public InputType types[4];
+	public int values[4];
+	public int axis_values[2];
+
+	public DPad () {
+		axis_values[0] = axis_values[1] = 0;
+		types[0] = types[1] = types[2] = types[3] = InputType.INVALID;
+	}
+}
+
+public class LibGamepad.Mapping : Object {
+	private InputType[] buttons_type;
+	private int[] buttons_value;
+	private InputType[] axes_type;
+	private int[] axes_value;
+	private DPad[] dpads;
+
+	public Mapping.from_sdl_string (string? mapping_string) throws MappingError {
+		if (mapping_string == null || mapping_string == "")
```

Comment:
This function is a bit long, what about splitting it up a bit? Maybe adding `parse_hat_mapping`, `parse_button_mapping` and `parse_axis_mapping` would help (just making this out of my head).